src/pub_verif_token.ts

```diff
@@ -247,10 +375,90 @@ export class Client {
             tokRes.blindSig,
             this.finData.inv,
         );
-        const token = new Token(BLIND_RSA, this.finData.authInput, authenticator);
+        const token = new Token(this.tokenType, this.finData.authInput, authenticator);
 
         this.finData = undefined;
 
         return token;
     }
 }
+
+export class Client extends PubliclyVerifiableClient {
+    constructor(mode: BlindRSAMode | PartiallyBlindRSAMode) {
+        super(mode);
+    }
+
+    async createTokenRequest(
+        tokChl: TokenChallenge,
+        issuerPublicKey: Uint8Array,
+    ): Promise<TokenRequest> {
+        return super._createTokenRequest(tokChl, issuerPublicKey);
+    }
+}
+
+export class ClientWithMetadata extends PubliclyVerifiableClient {
+    constructor(mode: BlindRSAMode | PartiallyBlindRSAMode, extensions: Extensions) {
+        super(mode, extensions);
+    }
+
+    async createTokenRequest(
+        tokChl: TokenChallenge,
+        issuerPublicKey: Uint8Array,
+    ): Promise<ExtendedTokenRequest> {
+        const tokenRequest = await super._createTokenRequest(tokChl, issuerPublicKey);
+        return new ExtendedTokenRequest(tokenRequest, this.extensions!);
+    }
+}
+
+abstract class PubliclyVerifiableOrigin {
+    private tokenType: BlindRSAType | PartiallyBlindRSAType;
+    private suite: Pick<BlindRSA, 'verify'>;
+
+    constructor(
+        public readonly mode: BlindRSAMode | PartiallyBlindRSAMode,
+        public readonly originInfo?: string[],
+        public readonly extensions?: Extensions,
+    ) {
+        if (this.extensions === undefined) {
+            this.suite = BLIND_RSA.suite[this.mode]();
+            this.tokenType = BLIND_RSA;
+        } else {
+            const suite = PARTIALLY_BLIND_RSA.suite[this.mode]();
+            const extensions = this.extensions.serialize();
+            this.suite = {
+                verify: (publicKey: CryptoKey, signature: Uint8Array, message: Uint8Array) =>
+                    suite.verify(publicKey, signature, message, extensions),
+            };
+            this.tokenType = PARTIALLY_BLIND_RSA;
```

Comment:
Current code can be simplified. What if we store both the suite and the extensions, then just call `verify` with the right parameters.

Reply:
the extension is stored. However, at the verification step, the type of the suite is not known. We therefore cannot easily pass `extensions` parameter at that time.
Having this function wrapping in the constructor, which stores the extension and passes them when necesarry allows to provide consistent typing.